chia/full_node/tx_processing_queue.py

```diff
@@ -0,0 +1,75 @@
+from __future__ import annotations
+
+import asyncio
+import logging
+from dataclasses import dataclass
+from typing import Dict, List, Optional
+
+from chia.server.ws_connection import WSChiaConnection
+from chia.types.transaction_queue_entry import TransactionQueueEntry
+
+
+@dataclass
+class TransactionQueue:
+    _list_iterator: int
+    _queue_length: asyncio.Semaphore
+    _index_to_peer_map: List[WSChiaConnection]
+    _queue_dict: Dict[WSChiaConnection, asyncio.Queue[TransactionQueueEntry]]
+    _high_priority_queue: asyncio.Queue[TransactionQueueEntry]
+    max_size: int
+    log: logging.Logger
+
+    def __init__(self, max_size: int, log: logging.Logger) -> None:
+        self._list_iterator = 0
+        self._queue_length = asyncio.Semaphore(0)  # default is 1
+        self._index_to_peer_map = []
+        self._queue_dict = {}
+        self._high_priority_queue = asyncio.Queue(max_size)
+        self.max_size = max_size
+        self.log = log
+
+    def qsize(self) -> int:
+        return self._queue_length._value
+
+    def full(self, peer: Optional[WSChiaConnection] = None) -> bool:
+        return self.qsize() == self.max_size or (peer is not None and not self._queue_dict[peer].full())
+
+    def empty(self) -> bool:
+        return self.qsize() == 0
+
+    async def put(self, tx: TransactionQueueEntry, high_priority: bool = False) -> None:
+        if tx.peer is None or high_priority:  # when it's local there is no peer.
+            await self._high_priority_queue.put(tx)
+        else:
+            if tx.peer not in self._queue_dict:
+                self._queue_dict[tx.peer] = asyncio.Queue(500)
+                self._index_to_peer_map.append(tx.peer)
+            try:
+                self._queue_dict[tx.peer].put_nowait(tx)
```

Comment:
did you consider making `put()` sleep here if the queue is full? That might simplify the API, you may no longer need `empty()`, `full()` or `qsize()`

Reply:
yeah i just changed it up and simplified the api